### PR TITLE
Exposing container logs URL in tony portal

### PR DESCRIPTION
**Pull request by pralabhkumar**
This PR is to enhance the tony portal(tony history server)  and helps  user to see the container logs .This functionality is very similar to Spark/MR history server , which provides a way for user to navigate container logs other than master . 

This PR 
- Add the container ID to the tony history file events 
- Add logs link to the tony history server . 
image
image

[Doc](https://docs.google.com/document/d/1NfXHHGh1kYRjZ4A5X71_LDxZNk0mVqcVvYzM_S8WsqA/edit#heading=h.sey16f3xv5ap) have more information about the same 

[Jira](https://jira01.corp.linkedin.com:8443/browse/DSS-24173) 

**Testing** 
Unit test cases written to check the functionality . E2E testing done on local machine

### CODE REVIEW

**Comment by oliverhu** [comment]
There should be a separate section on the job detail page to list all the logs, most events don't map to logs

ApplicationId 12345
----
Logs Table:
worker_1  container_log_1
worker_2  container_log_2
---
Events Table

**Comment by pralabhkumar** [comment]
> There should be a separate section on the job detail page to list all the logs, most events don't map to logs
> 
> ## ApplicationId 12345
> ## Logs Table:
> worker_1 container_log_1
> worker_2 container_log_2
> Events Table

@oliverhu 
APPLICATION_INTIATED & TASK_STARTED will have the logs , TASK_FINISHED & APPLICATION_FINISHED will not have (thats why there will be NA for them) . Since job event contains hostid and container (which is required to build log URL) , the link is added to JobEvent page . Also the mapping of taskIndex and hostid is also on job event page , it would be simpler for user to map log for failed task to container url , in the job event page. 

However if you think its better to have separate page , i'll do the change . Please let me know

**Comment by pralabhkumar** [comment]
@oliverhu  Please review the PR , have done the changes as suggested (please see the screenshot in the description above , for UI)

File: tony-core/src/main/avro/TaskStarted.avsc [hunk 1]
```diff
@@ -5,6 +5,7 @@
   "fields": [
     {"name": "taskType", "type": "string"},
     {"name": "taskIndex", "type": "int"},
-    {"name": "host", "type": "string"}
+    {"name": "host", "type": "string"},
+    {"name": "containerID", "type": ["null","string"], "default": null, "doc":  "Container  id"}
```
Comment (oliverhu) [inline tony-core/src/main/avro/TaskStarted.avsc]: I don't think containerId could be null ?

File: tony-core/src/main/avro/ApplicationInited.avsc [hunk 2]
```diff
@@ -5,6 +5,7 @@
   "fields": [
     {"name": "applicationId", "type": "string"},
     {"name": "numTasks", "type": "int"},
-    {"name": "host", "type": "string"}
+    {"name": "host", "type": "string"},
```
Comment (oliverhu) [inline tony-core/src/main/avro/ApplicationInited.avsc]: we can refactor [host, container id] into a separate scheme like ContainerInfo as this is used in both AM and workers

Comment (oliverhu) [reply tony-core/src/main/avro/ApplicationInited.avsc]: could be a follow up PR

File: tony-core/src/main/java/com/linkedin/tony/models/JobLog.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+
+/**
+ * JobLog is responsible for parsing the event and create the information
+ * about container logs .
+ */
+public class JobLog {
+  private static final Log LOG = LogFactory.getLog(JobLog.class);
+  private static final String LOG_URL_SCHEMA = "http://{0}/jobhistory/nmlogs/{1}:{2}/{3}/{4}/{5}";
+  private static final String MAPREDUCE_JOBHISTORY_WEBAPP_ADDRESS = "mapreduce.jobhistory.webapp.address";
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/models/JobLog.java]: why do you need MR history server ?

File: tony-core/src/main/java/com/linkedin/tony/models/JobLog.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+   * @param yarnConfiguration {@link YarnConfiguration} yarnconfiguration provided to TonY
+   * @return Pair contains mapreduce.jobhistory.webapp.address if able to get information from
+   * yarnconf ,other will return null
+   */
+  private static Pair<String, String> getJobHistoryNodeMgrAdd(YarnConfiguration yarnConfiguration) {
+    Pair<String, String> jobHistoryNodeMgrAddress = null;
+    if (isParameterValid(yarnConfiguration)) {
+      String mapReduceJobHistoryWebAddress = yarnConfiguration.get(MAPREDUCE_JOBHISTORY_WEBAPP_ADDRESS);
+      String yarnNodeManagerAddress = yarnConfiguration.get(YARN_NODEMANAGER_ADDRESS);
+      if (isParameterValid(mapReduceJobHistoryWebAddress, yarnNodeManagerAddress)) {
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/models/JobLog.java]: not getting this portion, I think you are trying to assemble the link to logs, why do you need MR web addess? YARN NM address should be enough

File: tony-portal/public/stylesheets/main.css [hunk 5]
```diff
@@ -9,6 +9,30 @@ html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockq
...
+li a {
+  display: block;
+  color: white;
+  text-align: center;
+  padding: 14px 16px;
+  text-decoration: none;
+}
+
+li a:hover {
+  background-color: #111;
```
Comment (oliverhu) [inline tony-portal/public/stylesheets/main.css]: change the color of the text when selected, in screenshot, the texts are greenish when selected, keep white should be ok

File: tony-core/src/main/avro/TaskStarted.avsc (hunk 1 above)
Comment (pralabhkumar) [reply tony-core/src/main/avro/TaskStarted.avsc]: done changes as suggested

File: tony-core/src/main/avro/ApplicationInited.avsc (hunk 2 above)
Comment (pralabhkumar) [reply tony-core/src/main/avro/ApplicationInited.avsc]: Yes will do in follow up PR

File: tony-core/src/main/java/com/linkedin/tony/models/JobLog.java (hunk 4 above)
Comment (pralabhkumar) [reply tony-core/src/main/java/com/linkedin/tony/models/JobLog.java]: Removed, changed the logic

File: tony-portal/public/stylesheets/main.css (hunk 5 above)
Comment (pralabhkumar) [reply tony-portal/public/stylesheets/main.css]: done

**Comment by pralabhkumar** [comment]
@oliverhu  done the changes as suggested . Please review

**Comment by oliverhu** [comment]
@pralabhkumar Thanks for your contribution!

**Comment by pralabhkumar** [comment]
@oliverhu  Thx for reviewing

**Pull request merged**